napari/animation/animate.py

```diff
@@ -0,0 +1,336 @@
+        self.set_to_keyframe(new_frame)
+
+    def key_back_frame(self, viewer):
+        """Go backwards in key-frame list"""
+
+        new_frame = (self.current_frame - 1) % len(self.key_frames)
+        self.set_to_keyframe(new_frame)
+
+    def set_to_keyframe(self, frame):
+        """Set the viewer to a given key-frame
+
+        Parameters
+        -------
+        frame : int
+            key-frame to visualize
+        """
+
+        self.current_frame = frame
+
+        for i in range(len(self.key_frames[frame]["sliders"])):
+            self.viewer.dims.set_point(i, self.key_frames[frame]["sliders"][i])
+
+        # set visibility of layers
+        for j in range(len(self.viewer.layers)):
+            self.viewer.layers[j].visible = self.key_frames[frame]["vis"][j]
+
+        # update state
+        self.viewer.dims.ndisplay = self.key_frames[frame]["ndisplay"]
+        self.viewer.window.qt_viewer.view.camera.set_state(
+            self.key_frames[frame]["camera"]
+        )
+        self.viewer.window.qt_viewer.view.camera.view_changed()
+
+    def create_state_dict(self):
+        """Create list of state dictionaries. For key-frames selected interactively,
+        add self.inter_steps emtpy frames between key-frames. For key-frames from scripts,
+        the number of empty frames ot add between each key-frame is already set in self.inter_steps.
+        """
+
+        if type(self.inter_steps) is not list:
+            inter_steps = len(self.key_frames) * [self.inter_steps]
+        else:
+            inter_steps = self.inter_steps
+
+        empty = {
+            "ndisplay": [],
+            "frame": [],
+            "camera": [],
+            "vis": [],
+            "sliders": [],
+        }
+        states_dict = []
+        for ind, x in enumerate(self.key_frames):
+            states_dict.append(x)
+            # do not add frames after last key-frame
+            if ind < len(self.key_frames) - 1:
+                # do not add frames when switching camera
+                if (
+                    self.key_frames[ind]['ndisplay']
+                    == self.key_frames[ind + 1]['ndisplay']
+                ):
+                    for y in range(inter_steps[ind]):
+                        states_dict.append(copy.deepcopy(empty))
+        for ind, x in enumerate(states_dict):
+            x["frame"] = ind
+        self.states_dict = states_dict
+
+    def create_steps(self):
+        """Interpolate states between key-frames"""
+
+        self.create_state_dict()
+        self.interpolated_states = util.interpolate(self.states_dict)
+
+    def key_interpolframe(self, viewer):
+        """Progress through interpolated frames"""
+
+        self.create_steps()
+
+        new_frame = (self.current_interpolframe + 1) % len(self.states_dict)
+        self.update_viewer_from_state(new_frame)
+        self.current_interpolframe = new_frame
+
+    def collect_images(self):
```

Comment:
Am I right in noting `collect_images` is no longer called, but you now do all the screenshooting inside `make_movie`? That's great, I think there should only be one function here that calls the screenshot method, but then `make_movie` should have the option to also just collect all the images into an array and return the array too.

So maybe if make `name=None` be the default argument to `make_movie` and `if name==None` then `make_movie` returns `image_stack` the 3D numpy array of screenshots without writing a file, if `make_movie` is actually writing a file I'd just return the path to the movie file. Curious what @jni thinks, but it is definitely nice that the movie writing now happens frame-by-frame, so if someone wants a movie that is 10000s of frames you don't have to create a numpy array with all that data in it in memory

Reply:
Very good idea. ```collect_images``` was just a left-over from previous iterations but I kept it around in case people need the stack, but what you propose removes a lot of redundancy.

Reply:
I'm not sure a 3D stack is a great idea, as it'll tend to be huge. I might prefer a *generator* of images that the user can then consume as they please — perhaps even using `np.fromiter` if they do want the giant array! I haven't looked at the rest of the code closely enough to know whether this is feasible...

Reply:
@jni it should be, but that maybe suggests a slightly different way of thinking about `collect_images` and `make_movie` and you might need to provide a little guidance on how that should be done. My current proposal was to just drop `collect_images` and put all the logic inside `make_movie` so that you didn't have to instantiate the full 3D array inside `make_movie`, but now with the idea of using a *generator* instead maybe we make the `collect_images` function return a generator, and then `make_movie` just does `np.fromiter` with the generator to save out each frame.

I actually like this idea the best as it means the `make_movie` logic is really only about the saving of the actual movie file, and the `collect_images` logic is only about the screenshoting and state updates of the viewer. Does this make sense @guiwitz?

Reply:
I now have a python generator function ```frame_generator()``` which creates all the frames as 2D numpy arrays. I left the option to generate a single specific frame if a ```frame``` argument is passed. The ```animate()``` function now only takes care of actually making the movie.